### PR TITLE
Fix finding of patterns and linetypes from measurement system.

### PR DESCRIPTION
**Pull request by dwsdolce**
The current code depends on the document unit instead of the measurement system - although the user interface says otherwise. In addition, the linetypes are initialized before the document has its variables defined. Fixed the CMakeLists.txt to be consisted with qmake. Fixed the plugin loaded to not expect _debug in dll plugin names - to be consistent with changes to both qmake and cmake.

### CODE REVIEW

**Comment by qcad** [comment]
Thank you for your report.
This has been partially manually merged and partially rejected.

**Comment by dwsdolce** [comment]
Can you let me know why it was partially rejected? As far as I can see you did not accept the change that makes what the code does agree with what the preference settings and documentation say the code does. Why is that?  Also, the CMakefile.txt changes are required to be able to build the community version on windows. Why was this rejected?

**Comment by qcad** [comment]
Drawing unit detection needs more thought as it breaks backwards compatibility. It needs either a separate configuration preference or detection of legacy drawing files. The cmake file change breaks things on our end. Can you elaborate on the problems you have experienced? We are building on all platforms, including Windows here.

**Comment by dwsdolce** [comment]
Hello Andrew,
Thank you for the response.  A couple of comments:


  1.  There was a change in CMakeInclude.txt and shared.pri in the root directory 7 months ago by you that removed the _debug decoration from the building of debug files. There was left code in src/core/RPluginLoader.cpp that has code in the routine RPluginLoader::getPluginFiles to check if you are running with QT_DEBUG set and then checking to make sure that _debug is present on the name of the file. So, if you build in debug mode you cannot run unless you manually rename all of the files to include _debug (since they no longer have that decoration when build). Seems a little ugly. So, I provided a change to remove the check since it is no longer necessary.
  2.  When I build opennurbs the dlls are defined to be dynamic libraries in the CMakeLists.txt for both opennurbs and zlib. The corresponding qmake files build them as static. The files you ship with the QCAD Pro snapshot are build as static libs. When built using CMake on windows the build breaks because it does not copy over both the .dll and .lib file and then fails to run. To make things consistent between what you ship, what QMake builds, what CMake builds I changed the CMakeLists.txt to be STATIC – that then works consistently. I do not know how you build on all platforms but I think there is an inconsistency in the checked in CMake, QMake files and what you release in the snapshots…
  3.  I understand the backward compatibility issue. The core issue is having the ability to specify the Measurement system as Metric/Imperial and the Drawing/Paper unit as different. I spent two days trying to figure out why putting my files in the imperial location failed by them being ignored. What I see in the code is that you are using RUnit::IsMetric to determine the Measurement system based on the underlying Units. This is good. What is confusing is the Text in the General: Drawing Unit form that says: “Measurement system (for line types and hatch patterns)”. So, maybe the correct change is to remove this comment and continue to use RUnit::IsMetric instead of RDocument::IsMetric. Clearly the information currently presented to the user is inconsistent with what the code does.

I am truly only trying to help. I spent most of my career (40 years) developing and managing EDA software (Analogy and Synopsys) development. I understand the need to maintain tight oversight on your code so that it does not fragment and become spaghetti code. You have been doing a great job in that and have an enviable position in that QCAD seems like the best opensource product for 2D cad available (I spent quite a lot of time going through other programs and trying to use them for a course on using CAD in drawing Guitar plans – making guitars being a retirement hobby). The course used RhinoCAD and QCAD was the only freely available tool I could find that could complete the coursework. I tried QCAD, FreeCAD, LibreCAD, Fusion 360, TurboCAD, OpenSCAD, SolidEdge Community, and Vectric VCarve (not free but I have a license). QCAD worked. This is what has motivated me to find solutions.

I would be interested in helping if that were possible.

So, in summary, I think the first two issues above are still a problem in what is checked in on the main branch. The 3rd issue is a problem but my solution may not be the correct one – changing the user interface or removing thee Measurement system option altogether may be a better solution.

Regards
David Smith
